textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/join/Join.java

```diff
@@ -42,14 +88,162 @@ public void open() throws Exception {
+		try {
+			outerOperator.close();
+			// innerOperator.close(); already called in open()
+
+		} catch(Exception e) {
+			e.printStackTrace();
+			throw new DataFlowException(e.getMessage(), e);
+		}
+		// Clear the inner tuple list from memory on close.
+		innerTupleList.clear();
+	}
+
+	// Used to compare IDs of the tuples.
+	private boolean compareId(ITuple outerTuple, ITuple innerTuple) {
+		// Check for the Validity of ID field and if both the ID fields are equal.
 		
+		// First check if the field in question exists by using try catch.
+		// Will throw an exception if it doesn't exist. This leads to return false.
+		// Then check if both the fields are of type IntegerField.
+		// --> This is the bare minimum thing that can be done to to verify valid 
+		// id attribute (as of now) (probably it is better to add a field called ID).
+		String fieldName = joinPredicate.getidAttribute().getFieldName();
+		try {
+			if(outerTuple.getField(fieldName).getClass().equals(IntegerField.class)&&
+					innerTuple.getField(fieldName).getClass().equals(IntegerField.class)) {
+				if(outerTuple.getField(fieldName).getValue()
+						.equals(innerTuple.getField(fieldName).getValue())) {
+					return true;
+				}
+			}
+		} catch(Exception e) {
+			;
+		}
+		return false;
+	}
+
+	// Process the tuples to get a tuple with join result if predicate is satisfied.
+	private ITuple joinTuples(ITuple outerTuple, ITuple innerTuple, JoinPredicate joinPredicate) throws Exception {
+		ITuple nextTuple = null;
+		List<Span> newJoinSpanList = new ArrayList<>();
+
+		if(!compareId(outerTuple, innerTuple)) {
+			return null;
+		}
+
+		// If either/both tuples have no span information, return null.
+		// Check using try/catch if both the tuples have span information.
+		// If not return null; so we can process next tuple.
+		
+		IField spanFieldOfInnerTuple = null;
+		IField spanFieldOfOuterTuple = null;
+		try {
+			spanFieldOfInnerTuple = innerTuple.getField(SchemaConstants.SPAN_LIST);
+			spanFieldOfOuterTuple = outerTuple.getField(SchemaConstants.SPAN_LIST);
+		} catch(Exception e) {
+			return null;
+		}
+		// If either/both tuples have no span information, return null.
+		//		if(indexOfInnerSpanList == null || indexOfOuterSpanList == null) {
+		//			return null;
+		//		}
+
+		List<Span> innerSpanList = null;
+		List<Span> outerSpanList = null;
+		// Check if both the fields obtained from the indexes are indeed of type ListField
+		// TODO(Flavio): should SchemaConstants.SPAN_LIST be a reserved or special 
+		// fieldName? In that case we don't need to check if the classes are equal
+		if(spanFieldOfInnerTuple.getClass().equals(ListField.class)) {
+			innerSpanList = (List<Span>) spanFieldOfInnerTuple.getValue();
+		}
+		if(spanFieldOfOuterTuple.getClass().equals(ListField.class)) {
+			outerSpanList = (List<Span>) spanFieldOfOuterTuple.getValue();
+		}
+
+		Iterator<Span> outerSpanIter = outerSpanList.iterator();
+
+		while(outerSpanIter.hasNext()) {
```

Comment:
We use two loops to join the two span lists of the pair of tuples.  In future, we can optimize it by sorting the spans based on the start position, and then doing a "sort merge" of the two lists.

If you agree, please add a "TODO" with this comment to the code.